Microsoft.Quantum.Canon/Utils/Multiplexor.qs

```diff
@@ -0,0 +1,285 @@
+    ///
+    /// # Input
+    /// ## coefficients
+    /// Array of up to $2^n$ coefficients $\theta_j$. The $j$th coefficient 
+    /// indexes the number state $\ket{j}$ encoded in big-endian format. 
+    ///
+    /// ## pauli
+    /// Pauli operator $P$ that determines axis of rotation.
+    ///
+    /// ## control
+    /// $n$-qubit control register that encodes number states $\ket{j}$ in
+    /// big-endian format.
+    ///
+    /// ## target
+    /// Single qubit register that is rotated by $e^{i P \theta_j}$.
+    ///
+    /// # Remarks
+    /// `coefficients` will be padded with elements $\theta_j = 0.0$ if 
+    /// fewer than $2^n$ are specified.
+    operation MultiplexorPauli(coefficients:Double[], pauli: Pauli, control: BigEndian, target: Qubit) : (){
+        body{
+            if(pauli == PauliZ){
+                let op = MultiplexorZ(coefficients, control, _);
+                op(target);
+            }
+            elif(pauli == PauliX){
+                let op = MultiplexorPauli(coefficients, PauliZ, control, _);
+                WithCA(H, op, target);
+            }
+            elif(pauli == PauliY){
+                let op = MultiplexorPauli(coefficients, PauliX, control, _);
+                WithCA(Adjoint S, op, target);
+            }
+            elif(pauli == PauliI){
+                DiagonalUnitary(coefficients, control);
+            }
+            else{
+                fail $"MultiplexorPauli failed. Invalid pauli {pauli}.";
+            }
+        }
+        adjoint auto
+        controlled auto
+        adjoint controlled auto
+    }
+
+    /// # Summary
+    /// Multiply-controlled unitary operation $U$ that performs rotations
+    /// by angle $\theta_j$ about single-qubit Pauli operator $Z$ when 
+    /// controlled by $n$-qubit number state $\ket{j}$.
+    ///
+    /// $U = \sum^{2^n-1}_{j=0}\ket{j}\bra{j}\otimes e^{i Z \theta_j}$.
+    ///
+    /// # Input
+    /// ## coefficients
+    /// Array of up to $2^n$ coefficients $\theta_j$. The $j$th coefficient 
+    /// indexes the number state $\ket{j}$ encoded in big-endian format. 
+    ///
+    /// ## control
+    /// $n$-qubit control register that encodes number states $\ket{j}$ in
+    /// big-endian format.
+    ///
+    /// ## target
+    /// Single qubit register that is rotated by $e^{i P \theta_j}$.
+    ///
+    /// # Remarks
+    /// `coefficients` will be padded with elements $\theta_j = 0.0$ if 
+    /// fewer than $2^n$ are specified.
+    ///
+    /// # References
+    /// - Synthesis of Quantum Logic Circuits
+    ///   Vivek V. Shende, Stephen S. Bullock, Igor L. Markov
+    ///   https://arxiv.org/abs/quant-ph/0406176
+    operation MultiplexorZ(coefficients: Double[], control: BigEndian, target: Qubit) : () {
+        body{
+            // pad coefficients length to a power of 2.
+            let maxCoefficients = 2^(Length(control));
+            let padZeros = new Double[maxCoefficients - Length(coefficients)];
```

Comment:
This might also be a reasonable function to extract into Arrays.qs. In particular, `Pad<'T> : ('T, Int, 'T[]) -> 'T[]`, such that this could be written as `Pad(0, maxCoefficients, coefficients)`.

Reply:
This Pad function will not work in some cases -- a runtime error is generated in cases where there is no default value of type 'T to pad with but the default value is called. Specifically, the default value of a new element of type (Qubit[] => () : ()) is undefined -- the use case here would require the NoOp.

Reply:
I got what you meant now -- the first 'T defines the default pad value.

Reply:
Added PadTail function.